### DISCUSSION TITLE
yolov10: anyone tried?

### DISCUSSION THREAD

**Discussion by lkaino** [blakeblackshear/frigate#11548] [General]
Yolov10 is out. Would be interesting to see how it performs on Frigate: https://github.com/THU-MIG/yolov10. Has the interface changed a lot?

**Comment by ivanjx** [comment]
the yolov5 and yolov8 have been removed from frigate 0.14.0 due to licensing issues. i dont think yolov10 can be added in the forseable future. for now i am trying out yolo-nas.

**Reply by lkaino** [reply]
Interesting, where is the removal documented? Is it only the models that are removed or also the integration code?

**Reply by ivanjx** [reply]
everything about yolo is removed so no more of it on 0.14 once it is released. more details in this commit: https://github.com/blakeblackshear/frigate/commit/14235c42b90ebcd67c8500086a67421f6c91826e

**Reply by ivanjx** [reply]
though, yolox is still there and they are adding yolo-nas as replacement here https://github.com/blakeblackshear/frigate/commit/4313fd97aaa308a2b60b65bfd3f6bb334237fc7b

**Reply by lkaino** [reply]
Thank you for this information! I'll try the nas version out.

**Comment by lkaino** [comment]
I configured the NAS version today. So far I'm really impressed with the performance. Accuracy is better compared to mobilenet I would say.

**Reply by philippedegeer** [reply]
Same thing here. Detection seems very good, especially in difficult cases, such as low contrast or when the motion box is much larger than the object to detect. In this area, it looks even better than Yolov8. I also had no FP in more than a week. I will continue testing and report accordingly.

**Reply by lkaino** [reply]
Which version of the model are you running? I'm using the small (s). It even detects dogs and cats which are really small in the camera image.

**Reply by ivanjx** [reply]
yes the reduced FP is the most obvious improvement of this model compared to v8.
btw im using the S model with size 736x736 with detect size 1280x736 and my camera is 1080p.

**Reply by philippedegeer** [reply]
I am using the s version with size 256x256 for now, because of my limited hardware capability. I get an inference speed of 48ms out of a Celeron J4125 iGPU. I am considering purchasing a Core Ultra 7 with an Arc iGPU  NUC to get more speed.

**Reply by ivanjx** [reply]
@philippedegeer what is your detect size? if i have my input size to be a lot smaller (like 320) and my camera is 1080p then it simply wont detect far objects despite the claim by frigate that it will cut required frame and send it to the detector. keeping the input and detect size at 736p fixed it for me.

**Reply by philippedegeer** [reply]
Currently my detect size is 800x450 but I have played a lot with it, even up to 1920x1080. This clearly has an influence on far objects detection. What I have now seems ok for my use case. I am easily detecting persons 12 meters away from a wide angle (horizontal 129°) Axis camera.

**Reply by ivanjx** [reply]
i guess my camera quality is too bad on low res that i need to use the 1080p stream and resize it to 736p and use the model with the same size. 110ms speed is definitely a very bad number but since i only have 2 cameras and one of them i set to 2 fps detection it is not a big problem for me.

**Comment by canadarob** [comment]
I trained my fine tuned model in yolov10 to compare to yolov8. There is a slight increase in inference time but the detection is impressive. Smaller objects are detected better and prediction scores are higher.

I even trained the coco dataset with relu6 activation. Then fine tuned on top. This halfed inference time the same way it did for yolov8. For edgetpu. 

Yolov8n=14.5ms , yolov10n= 16.5ms on edgetpu with 512x512

I also have it running on openvino. Same int8 results as edgetpu but faster.

**Reply by lkaino** [reply]
Have you tried the NAS model? I'm really impressed on its performance. It would be interesting to hear how it compares with v10.

**Reply by canadarob** [reply]
I have not tried NAS yet. As my use cases are mostly edge devices. I see that NAS currently doesn't have a Nano sized model. I don't have much experience scaling existing model architectures. Some of NAS ideas were used in yolov10. So it might have similar results for you.

**Reply by philippedegeer** [reply]
@canadarob Are you using yolov10 with Frigate 13.2 or with the 14beta version ?

**Reply by canadarob** [reply]
> @canadarob Are you using yolov10 with Frigate 13.2 or with the 14beta version ?

I'm running it on 13.2 but I have a local pull with some patches. Its based on what harakas did with their yolov8 updates and modified to run v10. Also i removed nms as its not required with the one to one head of yolov10 and frigates own confidence based selection.

**Reply by philippedegeer** [reply]
Thanks canadarob, I will give it a try.

**Reply by canadarob** [reply]
> Thanks canadarob, I will give it a try.

To note,  yolov10 outputs predicted boxes in xyxy not xywh like yolov8. That one wasted a few of my hours lol

**Reply by andrey-skat** [reply]
> To note, yolov10 outputs predicted boxes in xyxy not xywh like yolov8. That one wasted a few of my hours lol

@canadarob is it the same code as for yolonas? Can you push your fork? :)

**Reply by canadarob** [reply]
> > To note, yolov10 outputs predicted boxes in xyxy not xywh like yolov8. That one wasted a few of my hours lol
> 
> @canadarob is it the same code as for yolonas? Can you push your fork? :)

It's very similar to harakas's yolov8 integration. With nms modifications and changing the postprocess for xyxy output.
I can push a fork early next week. I am currently out on vacation with limited access.

**Reply by andrey-skat** [reply]
@canadarob just friendly reminder :)

**Reply by canadarob** [reply]
> @canadarob just friendly reminder :)

It appears there is a licensing issue between Ultralytics and Frigate that wont get resolved. So I rather not push a fork with that violation. I will upload the files that I modified to this forum asap. Including some instructions. Then you can privately make those modification on your own pull of frigate.

**Reply by canadarob** [reply]
[
[openvino.txt](https://github.com/user-attachments/files/16429861/openvino.txt)
[detector_config.txt](https://github.com/user-attachments/files/16429862/detector_config.txt)
[edgetpu_tfl.txt](https://github.com/user-attachments/files/16429863/edgetpu_tfl.txt)
[util.txt](https://github.com/user-attachments/files/16429864/util.txt)
](url)


Those are the files I used on frigate 13.2. The above .txt file of course will have to be changed to .py and sit in the directories as below:
/opt/frigate/frigate/detectors/plugins/edgetpu_tfl.py
/opt/frigate/frigate/detectors/util.py
/opt/frigate/frigate/detectors/plugins/openvino.py
/opt/frigate/frigate/detectors/detector_config.py

You can use yolov8 or yolov10. Frigate config example below for coral and openvino:
model:
  labelmap_path: /config/model_cache/yolov8/roblabelsc.txt
  model_type: yolov10
  height: 512
  width: 512
detectors:
  coral_pci0:
    type: edgetpu
    device: pci:0
    model:
      path: /config/model_cache/yolov8/v10relu6v3a_full_integer_quant_edgetpu.tflite
  coral_pci1:
    type: edgetpu
    device: pci:1
    model:
      path: /config/model_cache/yolov8/v10relu6v3a_full_integer_quant_edgetpu.tflite
  coral_pci2:
    type: edgetpu
    device: pci:2
    model:
      path: /config/model_cache/yolov8/v10relu6v3a_full_integer_quant_edgetpu.tflite
  openvino:
    type: openvino
    device: GPU
    model:
      path: /config/model_cache/yolov8/v10relu6v3_ser512.xml

Notes for v10:
- For both tflite conversion and openvino conversion the TopK has to be removed from the model. This is not an issue with the one2one head as it will only give a few bounding box outputs and we do our "own" topk in the postprocessing. See here: https://github.com/THU-MIG/yolov10/issues/72. You could also use this persons remove_topk code https://github.com/jayfalls/jetson_nano_ubuntu20_docker/blob/main/yolo/v10_fixer.py within your ultralytics package exporter.py.
- NMS is removed in v10 post process. It is no longer required.
- group convolution has to be disabled when exporting to tflite. Like  below in ultralytics exporter.py:
        onnx2tf.convert(
            input_onnx_file_path=f_onnx,
            output_folder_path=str(f),
            not_use_onnxsim=True,
            verbosity=verbosity,
            output_integer_quantized_tflite=self.args.int8,
            quant_type="per-tensor",  # "per-tensor" (faster) or "per-channel" (slower but more accurate)
            custom_input_op_name_np_data_path=np_data,
            disable_group_convolution=True,
        )
Notes for both v8 and v10:
- adding this argument below to the onnx2tf config will change the tensor io type to uint which can further decrease inference time on edgetpu's:
 input_output_quant_dtype="uint8", for both v10 and v8. Theres is a change to autobackend.py in ultralytics to properly validate uint8 model outputs. ill post that a bit later. 
- Frigate requires Openvino to be upgraded to the latest to allow for the new int8 quantization.

Results:
- I have only quick observations. But the small object detection is better with yolvo10 with my custom model. I have better validation results when outputting to edgetpu on v10 vs v8.  This is the same for openvino int8 outputs. I will continue trying both.
- Edgetpu inference at 512x512 nano model with normal silu activations is around 27ms on v8 and 30ms on v10. When trained on relu6 activation the inference goes down to 14.5ms on v8 and 16.7ms on v10. This is with minimal MAP loss. Also, This is using uint8 tensor io described above.
- I will post a quick video later on. As well as some validation results.
- I am exporting a v10n model for the edgetpu and openvino for you to easily try. Ill also export a v8 and compare the results at smaller image sizes.

**Reply by canadarob** [reply]
![Screenshot from 2024-07-30 13-05-40](https://github.com/user-attachments/assets/b6333ddb-2e08-46c8-b35e-2587dd8f535d)

Here is a relu6 yolov10n 512x512, coco base weights. You can see around 20ms:
https://github.com/canadarob/edgetpubases/blob/main/yolov10n-relu6_full_integer_quant_edgetpu.tflite

In addition here is a relu6 activated yolov10n 320x320, coco base weights. 11.6ms inference:
https://github.com/canadarob/edgetpubases/blob/main/yolov10n-relu6_full_integer_quant_edgetpu320.tflite

Labels file:
[labels.txt](https://github.com/user-attachments/files/16431437/labels.txt)

Ill try to get some normal silu activated yolov10n modes outputted later. They cannot map as many operations to the edgetpu. Openvino does both activation types well.

**Reply by Cheerpipe** [reply]
Thanks. Tested the 320 relu6 model and got a Nice 20+-2ms inference speed with a USB edgetpu.

50ms with the 512 version.

**Reply by canadarob** [reply]
https://github.com/canadarob/edgetpubases/blob/main/yolov10n512.tar

This is a openvino int8 model , yolov10n at 512x512 to try out.
This is the normal SILU activation. Openvino has much less slowdowns with this activation.

I will also provide a relu6 version as well

**Reply by andrey-skat** [reply]
Small offtop: didn't try yolov10 yet, but I have strange bug with yolov8 on newest openvino. When I export it from onnx with `ovc`, detector starts see all the time every label in every frame :) And I have to export to openvino with 2022.3 `mo ` to work it properly. Is there parameters I should add on train or export?

**Reply by canadarob** [reply]
> Small offtop: didn't try yolov10 yet, but I have strange bug with yolov8 on newest openvino. When I export it from onnx with `ovc`, detector starts see all the time every label in every frame :) And I have to export to openvino with 2022.3 `mo ` to work it properly. Is there parameters I should add on train or export?

you could try the export function within newest ultralytics package. See if that changes anything. If it works, you can then look at their exporter.py code to see what they did different with ovc. With yolov8 and latest ultralytics package I can export to openvino and use that with frigate. Are you using compress_to_fp16 option with ovc? Or using the standard float32 model

**Reply by andrey-skat** [reply]
No, standard export doesn't work for me. With ovc I tried both true and false for compress_to_fp16

I forgot to mention I do it on frigate 0.14 with ad-hoc patch that solves this error for me `[GPU] Mismatch tensor and port type: f32 vs u8` . Maybe that's the problem :)

**Reply by Taik** [reply]
@canadarob thanks for sharing the exported models --I was able to get `yolov10n-relu6_full_integer_quant_edgetpu320.tflite` running on my Coral USB ([with your patch](https://github.com/blakeblackshear/frigate/discussions/11548#discussioncomment-10192306)) - working nicely & way less false positives! 

> Thanks. Tested the 320 relu6 model and got a Nice 20+-2ms inference speed with a USB edgetpu.

I can confirm - seeing similar numbers (~25ms) with Coral USB.

**Reply by browndane** [reply]
Does the ad-hoc patch work for frigate 0.14? When I use the patch on 0.14, I get a weird error about path and labelmap_path. Works on the default model but not on yolov10. My detector part of my config looks like this.

detectors:
     coral:
     type: edgetpu
     device: usb
     model:
        path: /config/model_cache/yolov10/yolov10n-relu6_full_integer_quant_edgetpu320.tflite
        width: 320
        height: 320
        
The error is asking for path and labelpath (error stays the same when I add labelpath) for some reason:

frigate  | 2024-08-24 20:16:44.861515807  ***    Your config file is not valid!                     ***
frigate  | 2024-08-24 20:16:44.861518399  ***    Please check the docs at                           ***
frigate  | 2024-08-24 20:16:44.861520807  ***    https://docs.frigate.video/configuration/index     ***
frigate  | 2024-08-24 20:16:44.861662102  *************************************************************
frigate  | 2024-08-24 20:16:44.861666342  *************************************************************
frigate  | 2024-08-24 20:16:44.861668713  ***    Config Validation Errors                           ***
frigate  | 2024-08-24 20:16:44.861670972  *************************************************************
frigate  | 2024-08-24 20:16:44.861686139  path: Field required
frigate  | 2024-08-24 20:16:44.861702842  labelmap_path: Field required
frigate  | 2024-08-24 20:16:44.861720027  *************************************************************
frigate  | 2024-08-24 20:16:44.861723083  ***    End Config Validation Errors                       ***

**Reply by woloss** [reply]
Tried v10 on 0.14.1, for `person` detection it works really well, inference of edgetpu bumped from 20ms to ~30ms. Sadly not detecting `cat, dog` at all, even though I have `min_score: 0.4, threshold: 0.6`.

**Comment by mikezue-co** [comment]
@OrionFanWeb1701 odezwij się do mnie na maila, mamy duży zleceń dla Ciebie
